src/commands/mod.rs

```diff
@@ -2004,15 +2007,76 @@ fn cmd_new(ui: &mut Ui, command: &CommandHelper, args: &NewArgs) -> Result<(), C
         !args.revisions.is_empty(),
         "expected a non-empty list from clap"
     );
-    let commits = resolve_base_revs(&workspace_command, &args.revisions)?;
-    let parent_ids = commits.iter().map(|c| c.id().clone()).collect();
+    let target_commits = resolve_base_revs(&workspace_command, &args.revisions)?;
+    let target_ids = target_commits.iter().map(|c| c.id().clone()).collect_vec();
     let mut tx = workspace_command.start_transaction("new empty commit");
-    let merged_tree = merge_commit_trees(tx.base_repo().as_repo_ref(), &commits);
-    let new_commit = tx
-        .mut_repo()
-        .new_commit(command.settings(), parent_ids, merged_tree.id().clone())
-        .set_description(&args.message)
-        .write()?;
+    let mut num_rebased = 0;
+    let new_commit;
+    if args.insert_before {
+        // Instead of having the new commit as a child of the changes given on the
+        // command line, add it between the changes' parents and the changes.
+        // The parents of the new commit will be the parents of the target commits
+        // which are not descendants of other target commits.
+        let root_commit = tx.base_repo().store().root_commit();
+        if target_ids.contains(root_commit.id()) {
+            return Err(user_error("Cannot insert a commit before the root commit"));
+        }
+        let new_children = RevsetExpression::commits(target_ids.clone());
+        let new_parents = new_children.parents();
+        if let Some(commit_id) = tx
+            .base_workspace_helper()
+            .evaluate_revset(&new_children.dag_range_to(&new_parents))?
+            .iter()
+            .commit_ids()
+            .next()
+        {
+            return Err(user_error(format!(
+                "Refusing to create a loop: commit {} would be both an ancestor and a descendant \
+                 of the new commit",
+                short_commit_hash(&commit_id),
+            )));
+        }
+        let mut new_parents_commits: Vec<Commit> = tx
+            .base_workspace_helper()
+            .evaluate_revset(&new_parents)?
+            .iter()
+            .commits(tx.base_repo().store())
+            .try_collect()?;
+        if new_parents_commits.len() > 1 {
+            new_parents_commits.retain(|c| c != &root_commit);
```

Comment:
I didn't understand what this was about until I reviewed #1189. The problem is that the Git backend doesn't support creating merge commits that have the root commit as one parent, right? Could you add a comment about that?

Reply:
I sent #1204 to handle that better in the Git backend if the CLI (or other library user) were to forget this check.

Reply:
Comment added. Of course even with #1204 I have to keep this check since I don't want it to be an error here, I just remove the root commit from the set as it does exactly what we want.